models/classes/creator/TreeItemLookup.php

```diff
@@ -72,23 +73,30 @@ public function getItems(
         $data = $this->getTreeResourceLookupService()->getResources($itemClass, [], $propertyFilters, $offset, $limit);
 
         return $this->formatTreeData(
-            $this->filterTreeData($itemClass, $data)
+            $this->filterTreeData($data)
         );
     }
 
-    private function filterTreeData(core_kernel_classes_Class $root, array $treeData): array
+    private function filterTreeData(array $treeData): array
     {
         if (empty($treeData) || empty($treeData[0]['children'])) {
             return $treeData;
         }
 
-        /** @noinspection PhpUnhandledExceptionInspection */
-        $readableResourcesMap = $this->getSecureResourceService()->getAllChildren($root);
+        $resourceService = $this->getSecureResourceService();
 
         $treeData[0]['children'] = array_filter(
             $treeData[0]['children'],
-            static function (array $item) use ($readableResourcesMap): bool {
-                return $item['type'] !== 'instance' || isset($readableResourcesMap[$item['uri']]);
+            static function (array $item) use ($resourceService): bool {
+                try {
+                    if ($item['type'] === 'instance') {
+                        $resourceService->validatePermission($item['uri'], ['READ']);
```

Comment:
BTW, what do you think of using `PermissionHelper` in order to fetch all the child permissions at once?
In the following manner.
```php
$nodeIds = [];

foreach ($treeData[0]['children'] as $child) {
    if ($child['type'] === 'instance') {
        $nodeIds[] = $child['uri'];
    }
}

$accessibleNodes = $this->getServiceLocator()->get(PermissionHelper::class)
    ->filterByPermission(
        $nodeIds,
        PermissionInterface::RIGHT_READ
    );

$treeData[0]['children'] = array_filter(
    $treeData[0]['children'],
    static function (array $item) use ($accessibleNodes): bool {
        return $item['type'] !== 'instance' || in_array($item['uri'], $accessibleNodes, true);
    }
);
```

Reply:
didn't know about PermissionHelper
yep. it should work faster if we don't need cache.